pkg/storage/ingest/fetcher.go

```diff
@@ -496,6 +510,7 @@ func (r *concurrentFetchers) run(ctx context.Context, wants chan fetchWant, logg
 			attemptSpan.SetTag("attempt", attempt)
 
 			f := r.fetchSingle(ctx, w)
+			r.bufferedFetchedRecords.Add(int64(len(f.FetchPartition.Records)))
```

Comment:
if something calls `Stop()` before these records are pushed to `orderedFetches` then the atomic counter will always be positive. Should we decrement the counter if we give up on sending these records?

I'm thinking about the cases in this select which end up returning

```go
			select {
			case <-r.done:
				wantSpan.Finish()
				attemptSpan.Finish()
				close(w.result)
				return
			case w.result <- f:
				previousResult = fetchResult{}
			case <-ctx.Done():
			default:
				if w.startOffset >= w.endOffset {
					// We've fetched all we were asked for the whole batch is ready, and we definitely have to wait to send on the channel now.
					f.startWaitingForConsumption()
					select {
					case <-r.done:
						wantSpan.Finish()
						attemptSpan.Finish()
						close(w.result)
						return
					case w.result <- f:
						previousResult = fetchResult{}
					case <-ctx.Done():
					}
				}
			}
```

at this point you can also try to unify all the places which do cleanup actions

Reply:
> if something calls Stop() before these records are pushed to orderedFetches then the atomic counter will always be positive

I thought about it too, but do we really care?

- If the stop is called at shutdown, we don't care.
- If the stop is called because we're moving from concurrent fetcher at startup to sequential fetchwe when ongoing, we don't care because the fetcher reference is trashed
- If the stop is called because we're updating the concurrent fetcher config (`Update()`) then we care... so what if we simply reset the buffer in `Stop()` after the `r.wg.Wait()` to keep it simple?

Reply:
> * If the stop is called because we're updating the concurrent fetcher config (`Update()`) then we care... so what if we simply reset the buffer in `Stop()` after the `r.wg.Wait()` to keep it simple?

yeah that's another option 👍

Reply:
I'm adding some more unit tests and I found an issue. The way we currently increase the number of records is wrong because the same record may be concurrently fetched by multiple routines. I'm working on a fix.

Reply:
> I'm adding some more unit tests and I found an issue. The way we currently increase the number of records is wrong because the same record may be concurrently fetched by multiple routines. I'm working on a fix.

Not true. It was my test that was not doing the assertion correctly. I'm getting back to the Josh implementation.